src/mito2/src/memtable/merge_tree/partition.rs

```diff
@@ -108,22 +108,30 @@ impl Partition {
 
     /// Scans data in the partition.
     pub fn read(&self, mut context: ReadPartitionContext) -> Result<PartitionReader> {
-        let nodes = {
+        let (builder_source, mut nodes) = {
             let inner = self.inner.read().unwrap();
-            let mut nodes = Vec::with_capacity(inner.shards.len() + 1);
-            if !inner.shard_builder.is_empty() {
+            let mut shard_source = Vec::with_capacity(inner.shards.len() + 1);
+            let builder_reader = if !inner.shard_builder.is_empty() {
                 let builder_reader = inner.shard_builder.read(&mut context.pk_weights)?;
-                nodes.push(ShardNode::new(ShardSource::Builder(builder_reader)));
-            }
+                Some(builder_reader)
+            } else {
+                None
+            };
             for shard in &inner.shards {
                 if !shard.is_empty() {
                     let shard_reader = shard.read()?;
```

Comment:
We might also build the shard reader outside of the lock

Reply:
> We might also build the shard reader outside of the lock

fixed here: https://github.com/GreptimeTeam/greptimedb/pull/3406/commits/97d3d453ad7e10825e6b0241e60762d130f42b50